contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/azkaban/GobblinDistcpJob.java

```diff
@@ -0,0 +1,197 @@
+
+        if (!cdnURL.isEmpty()) {
+            info("####################################");
+            info("              Distcp");
+            info("####################################");
+            info("Using CDN: " + cdnURL);
+            try {
+                String username = props.getString("env.USER", "unknownClient");
+                String pathPrefix = props.getString(VoldemortBuildAndPushJob.PUSH_CDN_PREFIX);
+                pathPrefix = pathPrefix.endsWith("/") ? pathPrefix : pathPrefix + "/";
+                // Replace original cluster with CDN, e.g. hdfs://original:9000/a/b/c => webhdfs://cdn:50070/a/b/c
+                String cdnDir = source.replaceAll(".*://.*?(?=/)", cdnURL + pathPrefix + username);
+                FileSystem cdnRootFS = getRootFS(cdnDir);
+                Path from = new Path(source);
+                Path to = new Path(cdnDir);
+
+                if (!prereqSatisfied(cdnURL)) {
+                    warn("Please add/append \"" + cdnURL + "\" to the \"other_namenodes\" attribute in your job specification.");
+                    throw new RuntimeException("\"other_namenodes\" does not contain the CDN cluster address " + cdnURL);
+                }
+
+                deleteDir(cdnRootFS, cdnDir);
+                runDistcp(from, to);
+                deleteDirOnExit(cdnRootFS, cdnDir);
+                addPermissionsToParents(cdnRootFS, cdnDir);
+                source = cdnDir;
+                info("Use CDN HDFS cluster: " + source);
+            } catch (Exception e) {
+                warn("An exception occurred during distcp: " + e.getMessage());
+                warn("Use original HDFS cluster: " + source);
+            }
+            info("####################################");
+            info("          End of Distcp");
+            info("####################################");
+        }
+    }
+
+    private void runDistcp(Path from, Path to) throws Exception {
+        info("source: " + from);
+        info("CDN: " + to);
+        EmbeddedGobblin embeddedGobblin = new EmbeddedGobblinDistcp(from, to).mrMode();
+
+        // Used for global throttling"
+        embeddedGobblin.distributeJar("lib/*");
+
+        for (Map.Entry<String, String> entry : this.props.entrySet()) {
+            if (entry.getKey() != null && (entry.getKey()).startsWith(ATTR_PREFIX)) {
+                String key = (entry.getKey()).substring(ATTR_PREFIX.length());
+                embeddedGobblin.setConfiguration(key, entry.getValue());
+            }
+        }
+        JobExecutionResult result =  embeddedGobblin.run();
+
+        if (!result.isSuccessful()) {
+            throw new RuntimeException("Distcp job failed!", result.getErrorCause());
+        }
+    }
+
+    private FileSystem getRootFS(String target) throws Exception {
+        // Remove directory path from URL, e.g. hdfs://hostname:9000/a/b/c => hdfs://hostname:9000
+        URI rootURI = new URI(target.replaceAll("(?<=:[0-9]{1,5})/.*", ""));
```

Comment:
Hmm.... I'm not sure I understand your regex-fu here. Are you confident this is going to catch all URI syntaxes properly?

Are there any safeguards we can put in place to detect if this doesn't work properly?

(i.e.: will `FileSystem.get(...)` throw an exception if it's passed a bogus `rootURI`? If yes, can we catch it and add a proper error message of our own?)